internal/encryption/keyprovider/openbao/config.go

```diff
@@ -0,0 +1,97 @@
+package openbao
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	openbao "github.com/openbao/openbao/api"
+	"github.com/opentofu/opentofu/internal/encryption/keyprovider"
+)
+
+type Config struct {
+	Address string `hcl:"address,optional"`
+	Token   string `hcl:"token,optional"`
+
+	KeyName   string `hcl:"key_name"`
+	KeyLength int    `hcl:"key_length,optional"`
+}
+
+func (c Config) Build() (keyprovider.KeyProvider, keyprovider.KeyMeta, error) {
+	if c.KeyName == "" {
+		return nil, nil, &keyprovider.ErrInvalidConfiguration{
+			Message: "no key name found",
+		}
+	}
+
+	if c.KeyLength == 0 {
+		c.KeyLength = defaultKeyLength
+	}
+
+	if err := validateKeyLength(c.KeyLength); err != nil {
+		return nil, nil, &keyprovider.ErrInvalidConfiguration{
+			Cause: err,
+		}
+	}
+
+	// DefaultConfig reads BAO_ADDR and some other optional env variables.
+	config := openbao.DefaultConfig()
+	if config.Error != nil {
+		return nil, nil, &keyprovider.ErrInvalidConfiguration{
+			Cause: config.Error,
+		}
+	}
+
+	// Address from HCL supersedes BAO_ADDR.
+	if c.Address != "" {
+		config.Address = c.Address
+	}
+
+	client, err := newClient(config, c.Token)
+	if err != nil {
+		return nil, nil, &keyprovider.ErrInvalidConfiguration{
```

Comment:
I think so. It fails if there's a problem parsing env variables so the error message looks like "failed to parse ENV_VAR_NAME".

However, we may fail to read env vars we don't use. Example: user sets valid address in config and invalid in BAO_ADDR -> tofu will fail because BAO_ADDR is invalid. I would not fix it because I believe it's not practically useful. WDYT?

Reply:
I think we may want to add a prefix, such as "failed to initialize the OpenBao key provider: whatever OpenBao said".

Reply:
`newClient` is our wrapper over openbao client constructor so the prefix is already present.